subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeBuildClassPathInitializer.java

```diff
@@ -46,13 +60,29 @@ public void execute(Configuration classpath) {
             tasksToBuild.add(TaskIdentifier.of(targetBuild.getBuildIdentifier(), (TaskInternal) task));
         }
         if (!tasksToBuild.isEmpty()) {
-            buildTreeWorkGraphController.withNewWorkGraph(graph -> {
-                graph
-                    .scheduleWork(builder -> builder.scheduleTasks(tasksToBuild))
-                    .runWork()
-                    .rethrow();
-                return null;
-            });
+            buildTreeWorkGraphController.withNewWorkGraph(graph -> buildBuildLogic(graph, tasksToBuild));
+        }
+    }
+
+    @Nullable
+    @SuppressWarnings("try")
+    private Void buildBuildLogic(BuildTreeWorkGraph graph, List<TaskIdentifier.TaskBasedTaskIdentifier> tasksToBuild) {
+        BuildTreeWorkGraph.FinalizedGraph finalizedGraph = graph.scheduleWork(builder -> builder.scheduleTasks(tasksToBuild));
+        try (FileLock ignored = buildLogicBuildLockFor(currentBuild)) {
+            finalizedGraph
+                .runWork()
+                .rethrow();
+            return null;
         }
     }
+
+    private FileLock buildLogicBuildLockFor(BuildState build) {
+        return fileLockManager.lock(
+            new File(build.getBuildRootDir(), ".gradle/noVersion/buildLogic"),
```

Comment:
💭 I wonder if this is the best name for the lock. If this is actually something `buildSrc` specific, I wouldn't go for another generic name, but rather for `buildSrcXxx`.

Could you also explain and/or add a comment on how these three things are connected?
- Composite build class-path
- buildSrc
- buildLogic